### DISCUSSION TITLE
Generate Mesh with different material blocks

### DISCUSSION THREAD

**Discussion by mfitzka** [idaholab/moose#23481] [Q&A Meshing] [answered]
Hi all,

Which internal mesh generator would be the most suitable for me?

I am looking to generate simple meshes at first with two different materials, which would be similar to a cylindrical pipe with steel on the outside and another material inside. 
This would be a first tests before I go on to more "advanced" geometries.

Is it possible to do this with something like GeneratedMesh directly in MOOSE or should I use something else like GMSH?

Thanks!

**Comment by GiudGiud** [comment, marked as answer]
Hello

Material blocks are subdomains or blocks in MOOSE.
Many generators allow for this feature. I think GeneratedMeshGenerator allows it, if not CartesianMeshGenerator is a good option

Guillaume

**Comment by mfitzka** [comment]
The CartesianMeshGenerator worked to define the materials to subdomains. 

However, now I am running into a new problem:
**ERROR**
Material properties must be retrieved during object construction. This is a code problem.

Does this also have to do with the generated mesh and the assignment of materials and its properties during the generation of the mesh?

**Reply by GiudGiud** [reply]
No this seems independent. Do you mind sharing your input file here?
Are you using a moose module or an app?

**Reply by mfitzka** [reply]
The simulation uses quite alot of custom built kernels, but I can also include them in the repository.

https://github.com/mfitzka/phoenixpublic

The input file is called SimpleTestCartesian.i

**Reply by GiudGiud** [reply]
ok then in one of the custom-made objects (or kernels) you must have something like `getMaterialProperty` outside of the constructor. This will not work, you need to modify the code to make it retrieve a *reference* to the material property during the construction of the kernel.

An example:
https://github.com/idaholab/moose/blob/next/framework/src/kernels/ADMatReaction.C
https://github.com/idaholab/moose/blob/next/framework/include/kernels/ADMatReaction.h

look at the _reaction_rate material property

**Reply by mfitzka** [reply]
I added the custom kernels used in the application in the repository. 

At first glance, I cant seem to find the problem, as some of them are original MOOSE code and in all of the others the getMaterialProperty command should be inside the constructor.

Am I overlooking something?

**Reply by GiudGiud** [reply]
I dont see any kernels in the repo.
Is it this one?
https://github.com/mfitzka/phoenixpublic

**Reply by mfitzka** [reply]
Sorry for the confusion, I forgot to confirm the uploads.
All of them can now be found in the folder "kernels".

**Reply by GiudGiud** [reply]
This kernel for example
https://github.com/mfitzka/phoenixpublic/blob/main/kernels/CoupledCureTimeDerivative.C

This should be in the constructor not the initialSetup
```
    _density = &getMaterialProperty<Real>("density");
    _Hr = &getMaterialProperty<Real>("Hr");
```

**Reply by mfitzka** [reply]
Simply moving it to the constructor did not work, as i got several errors.
(error expected "{" before "_density")

I guess I must also change the syntax to match the others, but with for example
_density(getMaterialProperty<Real>("density")), I also get errors.

Are there some tutorials regarding coding this, as I am completely new to C++?

**Reply by GiudGiud** [reply]
just do this:
```
    CoupledCureTimeDerivative::CoupledCureTimeDerivative(const InputParameters & parameters) :
        Kernel(parameters),
    _density(&getMaterialProperty<Real>("density")),
    _Hr(&getMaterialProperty<Real>("Hr")),
	_v_dot(coupledDot("v")),
	 _dv_dot(coupledDotDu("v")),
        _v_var(coupled("v"))
    {}
```

**Reply by mfitzka** [reply]
I tried this, but then I get the error: 
multiple initializations given for "CoupledCureTimeDerivative::_density" 
and the same for _Hr.

Should I remove the _density(NULL), and _Hr(NULL), lines?

**Reply by GiudGiud** [reply]
oh yes my bad

**Reply by mfitzka** [reply]
With them removed it now compiled without any errors and this can be closed as solved.

Thank you very much Guillaume!
